docs/notebooks/research/recursive_least_squares.ipynb

```diff
@@ -0,0 +1,287 @@
+    "\n",
+    "from nengolib import Lowpass\n",
+    "\n",
+    "from nengo.builder import Builder, Operator, Signal\n",
+    "from nengo.builder.operator import Reset\n",
+    "from nengo.learning_rules import LearningRuleType\n",
+    "from nengo.params import NumberParam, Default\n",
+    "from nengo.synapses import SynapseParam\n",
+    "\n",
+    "\n",
+    "class RLS(LearningRuleType):\n",
+    "\n",
+    "    modifies = 'decoders'\n",
+    "    probeable = ('pre_filtered', 'error', 'delta', 'inv_gamma')\n",
+    "\n",
+    "    pre_synapse = SynapseParam(\n",
+    "        'pre_synapse', readonly=True)\n",
+    "\n",
+    "    def __init__(self, learning_rate=1.0, pre_synapse=Lowpass(tau=0.005)):\n",
+    "        super(RLS, self).__init__(learning_rate, size_in='post_state')\n",
+    "        self.pre_synapse = pre_synapse\n",
+    "\n",
+    "    @property\n",
+    "    def _argreprs(self):\n",
+    "        args = []\n",
+    "        if self.learning_rate != 1.0:\n",
+    "            args.append(\"learning_rate=%g\" % self.learning_rate)\n",
+    "        if self.pre_synapse is not None and self.pre_synapse != Lowpass(tau=0.005):\n",
+    "            args.append(\"pre_synapse=%r\" % self.pre_synapse)\n",
+    "        return args\n",
+    "\n",
+    "\n",
+    "def build_or_passthrough(model, obj, signal):\n",
+    "    \"\"\"Builds the obj on signal, or returns the signal if obj is None.\"\"\"\n",
+    "    if obj is not None:\n",
+    "        signal = model.build(obj, signal)\n",
+    "    return signal\n",
+    "\n",
+    "\n",
+    "class SimRLS(Operator):\n",
+    "\n",
+    "    def __init__(self, pre_filtered, error, delta, inv_gamma, tag=None):\n",
+    "        super(SimRLS, self).__init__(tag=tag)\n",
+    "\n",
+    "        self.sets = []\n",
+    "        self.incs = []\n",
+    "        self.reads = [pre_filtered, error]\n",
+    "        self.updates = [delta, inv_gamma]\n",
+    "\n",
+    "    @property\n",
+    "    def delta(self):\n",
+    "        return self.updates[0]\n",
+    "\n",
+    "    @property\n",
+    "    def inv_gamma(self):\n",
+    "        return self.updates[1]\n",
+    "\n",
+    "    @property\n",
+    "    def pre_filtered(self):\n",
+    "        return self.reads[0]\n",
+    "\n",
+    "    @property\n",
+    "    def error(self):\n",
+    "        return self.reads[1]\n",
+    "    \n",
+    "    def _descstr(self):\n",
+    "        return 'pre=%s > %s' % (self.pre_filtered, self.delta)\n",
+    "\n",
+    "    def make_step(self, signals, dt, rng):\n",
+    "        r = signals[self.pre_filtered]\n",
+    "        delta = signals[self.delta]\n",
+    "        error = signals[self.error]\n",
+    "        P = signals[self.inv_gamma]\n",
+    "        \n",
+    "        def step_simbcm():\n",
+    "            # Note: dt is not used in learning rule\n",
+    "            rP = r.T.dot(P)\n",
+    "            P[...] -= P.dot(np.outer(r, rP)) / (1 + rP.dot(r))\n",
```

Comment:
Note to self: `P[...] -= np.outer(P.dot(r), rP) / (1 + rP.dot(r))` should be more efficient since it avoids a matrix-multiply.